packages/devtools_app/lib/src/screens/profiler/cpu_profile_model.dart

```diff
@@ -250,20 +261,138 @@ class CpuProfileData {
         stackDepth: superProfile.profileMetaData.stackDepth,
         time: subTimeRange,
       ),
+      rootedAtTags: superProfile.rootedAtTags,
+    );
+  }
+
+  /// Generate a cpu profile from [originalData] where the profile is broken
+  /// down for each tag of the given [type].
+  ///
+  /// [originalData] does not need to be [processed] to run this operation.
+  factory CpuProfileData.withTagRoots(
+    CpuProfileData originalData,
+    CpuProfilerTagType type,
+  ) {
+    final useUserTags = type == CpuProfilerTagType.user;
+    final tags = <String>{
+      for (final sample in originalData.cpuSamples)
+        useUserTags ? sample.userTag! : sample.vmTag!,
+    };
+
+    final tagProfiles = <String, CpuProfileData>{
+      for (final tag in tags)
+        tag: CpuProfileData._fromTag(originalData, tag, type),
+    };
+
+    final metaData = originalData.profileMetaData.copyWith();
+
+    // Use a SplayTreeMap so that map iteration will be in sorted key order.
+    // This keeps the visualization of the profile as consistent as possible
+    // when applying filters.
+    final SplayTreeMap<String, CpuStackFrame> stackFrames =
+        SplayTreeMap(stackFrameIdCompare);
+
+    final samples = <CpuSampleEvent>[];
+
+    int nextId = 1;
+
+    for (final tagProfileEntry in tagProfiles.entries) {
+      final tag = tagProfileEntry.key;
+      final tagProfile = tagProfileEntry.value;
+      if (tagProfile.cpuSamples.isEmpty) {
+        continue;
+      }
+      final isolateId = tagProfile.cpuSamples.first.leafId.split('-').first;
+      final tagId = '$isolateId-${nextId++}';
```

Comment:
could this stack frame id already be in use? this could cause collisions in the map.

Reply:
No, we're regenerating stack frame IDs for all the sub-profiles using this single counter, so IDs will remain unique.